src/util/daemon.c

```diff
@@ -0,0 +1,194 @@
+/* 
+ * Copyright (c) 2017, SingularityWare, LLC. All rights reserved.
+ * 
+ * This software is licensed under a 3-clause BSD license.  Please
+ * consult LICENSE file distributed with the sources of this project regarding
+ * your rights to use or distribute this software.
+ * 
+ */
+
+
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <fcntl.h>
+
+#include "config.h"
+#include "util/file.h"
+#include "util/util.h"
+#include "util/daemon.h"
+#include "util/registry.h"
+#include "util/message.c"
+#include "lib/image/image.h"
+#include "lib/runtime/runtime.h"
+#include "util/privilege.h"
+
+void daemon_file_parse(void) {
+    singularity_message(DEBUG, "reached file parse\n");
+    char *key, *val;
+    char *line = (char *)malloc(2048 * sizeof(char *));
+    FILE *file = fopen(singularity_registry_get("DAEMON_FILE"), "r");
+
+    while( fgets(line, 2048, file) ) {
+        key = strtok(line, "=\n");
+        val = strtok(NULL, "=\n");
+        singularity_message(DEBUG, "Read key-val pair %s=%s\n", key, val);
+        singularity_registry_set(key, val);
+    }
+}
+
+void daemon_file_write(int fd, char *key, char *val) {
+    int retval = 0;
+    errno = 0;
+    
+    singularity_message(DEBUG, "Called daemon_file_write(%d, %s, %s)\n", fd, key, val);
+    retval += write(fd, key, strlength(key, 2048));
+    retval += write(fd, "=", 1);
+    retval += write(fd, val, strlength(val, 2048));
+    retval += write(fd, "\n", 1);
+
+    if ( errno != 0 ) {
+        singularity_message(ERROR, "Unable to write to daemon file: %s\n", strerror(errno));
+        ABORT(255);
+    }
+}
+
+void singularity_daemon_path(void) {
+    char *daemon_path = (char *)malloc(2048 * sizeof(char));
+
+    int uid = singularity_priv_getuid();
+    char *dev_ino = file_devino(singularity_registry_get("IMAGE"));
+    char *name = singularity_registry_get("DAEMON_NAME");
+    
+    snprintf(daemon_path, 2048, "/tmp/.singularity-daemon-%d/%s-%s", uid, dev_ino, name);
```

Comment:
This could be safely an environment variable, but it runs the chance that if it is changed with daemons running, they may not be seen via `stop` or `list`. But because this is a simple ASCII based PID file, I'm not sure there is much need to move it.